Zend/zend_execute.c

```diff
@@ -1795,6 +1795,18 @@ static zend_never_inline void zend_fetch_dimension_address_UNSET(zval *result, z
 	zend_fetch_dimension_address(result, container_ptr, dim, dim_type, BP_VAR_UNSET);
 }
 
+static zend_always_inline int zend_fetch_dimension_address_read_valid_container(zval *container, int support_strings) {
+	if (EG(current_execute_data)->opline->opcode != ZEND_FETCH_LIST) {
+		if (EG(current_execute_data)->opline->op1_type != IS_VAR) {
```

Comment:
Yeah.  I looked into using ZVAL_ERROR(), however, it's at the end of `zend_fetch_dimension_address_read` that it sets the return to NULL if we hit error cases.  This could have some cascading effects which I notices, for example:
```
$a = null;
var_dump($a[0]);
```
would not be `NULL`, rather it'd be `UNDEFINED:0`.  This causes many many problems in existing tests where it's expecting the result of a failed lookup to be NULL.  I assume internally they are functionally similar but I haven't looked much deeper.

My thought to accommodate your example would be to look to the previous opcode and determine if a null in this context is error-able or not.  So if we're not looking at a compiled-var as the container, if it's null, but the previous op-code wasn't a ZEND_FETCH_DIM_R we should error.  This would limit the erroring on a null -only- to if the container wasn't a result of a fetch already.